dart/collision/fcl/FCLCollisionDetector.cpp

```diff
@@ -37,231 +37,1344 @@
+
+  assert(_mesh);
+  fcl::BVHModel<BV>* model = new fcl::BVHModel<BV>;
+  model->beginModel();
+
+  for (size_t i = 0; i < _mesh->mNumFaces; i++)
+  {
+    fcl::Vec3f vertices[3];
+    for (size_t j = 0; j < 3; j++)
+    {
+      const aiVector3D& vertex = _mesh->mVertices[_mesh->mFaces[i].mIndices[j]];
+      vertices[j] = fcl::Vec3f(vertex.x, vertex.y, vertex.z);
+    }
+    model->addTriangle(vertices[0], vertices[1], vertices[2]);
+  }
+
+  model->endModel();
+  return model;
+}
+
+} // anonymous namespace
+
+
+
+//==============================================================================
+std::shared_ptr<FCLCollisionDetector> FCLCollisionDetector::create()
+{
+  return std::shared_ptr<FCLCollisionDetector>(new FCLCollisionDetector());
+}
+
+//==============================================================================
+FCLCollisionDetector::~FCLCollisionDetector()
+{
+  assert(mShapeMap.empty());
+}
+
+//==============================================================================
+const std::string& FCLCollisionDetector::getTypeStatic()
+{
+  static const std::string& type("FCL");
+
+  return type;
+}
+
+//==============================================================================
+const std::string& FCLCollisionDetector::getType() const
+{
+  return FCLCollisionDetector::getTypeStatic();
+}
+
+//==============================================================================
+std::unique_ptr<CollisionGroup>
+FCLCollisionDetector::createCollisionGroup()
+{
+  return common::make_unique<FCLCollisionGroup>(shared_from_this());
+}
+
+//==============================================================================
+bool FCLCollisionDetector::detect(
+    CollisionGroup* group, const Option& option, Result& result)
+{
+  result.clear();
+
+  if (!group)
+    return false;
+
+  if (group->getCollisionDetector()->getType()
+      != FCLCollisionDetector::getTypeStatic())
+  {
+    return false;
+  }
+
+  group->updateEngineData();
+
+  auto castedData = static_cast<FCLCollisionGroup*>(group);
+  auto broadPhaseAlg = castedData->getFCLCollisionManager();
+
+  FCLCollisionCallbackData collData(
+        option, &result, mPrimitiveShapeType,
+        mContactPointComputationMethod);
+  broadPhaseAlg->collide(&collData, collisionCallback);
 
-  if (!request.enable_cost
-      && (result.isCollision())
-      && (result.numContacts() >= request.num_max_contacts))
+  return result.isCollision();
+}
+
+//==============================================================================
+bool FCLCollisionDetector::detect(
+    CollisionGroup* group1, CollisionGroup* group2,
+    const Option& option, Result& result)
+{
+  result.clear();
+
+  if (!group1 || !group2)
+    return false;
+
+  if (group1->getCollisionDetector()->getType()
+      != FCLCollisionDetector::getTypeStatic())
+  {
+      return false;
+  }
+
+  if (group2->getCollisionDetector()->getType()
+      != FCLCollisionDetector::getTypeStatic())
+  {
+      return false;
+  }
```

Comment:
It will work for now when it wasn't intended, though. However, this is prone to error. I think this should be more strict so that all the groups should be created from this collision detector as:

```cpp
  if ((this != group1->getCollisionDetector()) 
      || (this != group2->getCollisionDetector()))
  {
      return false;
  }
```